modules/transformation.py

```diff
@@ -0,0 +1,609 @@
+import datetime
+import json
+import logging
+import os
+import re
+from json import JSONDecodeError
+
+import apache_beam as beam
+from apache_beam.io import ReadFromPubSub, WriteToBigQuery, BigQueryDisposition
+import apache_beam.io.fileio as fileio
+from apache_beam.io.gcp.bigquery_tools import FileFormat, RetryStrategy
+from dateutil import parser as date_parser
+
+from modules import constants, utils
+
+
+class ReadHarFiles(beam.PTransform):
+    def __init__(self, subscription=None, _input=None):
+        super().__init__()
+        self.subscription = subscription
+        self.input = _input
+
+    def expand(self, p):
+        # PubSub pipeline
+        if self.subscription:
+            files = (
+                p
+                | ReadFromPubSub(subscription=self.subscription, with_attributes=True)
+                | beam.Filter(lambda e: e.attributes["objectId"].endswith(".har.gz"))
+                | "GetFileName"
+                >> beam.Map(
+                    lambda e: f"gs://{e.attributes['bucketId']}/{e.attributes['objectId']}"
+                )
+                | beam.io.ReadAllFromText(with_filename=True)
+            )
+        # GCS pipeline
+        else:
+            matching = (
+                self.input if ".har.gz" in self.input else f"{self.input}/*.har.gz"
+            )
+
+            # TODO using ReadAllFromText instead of ReadFromText[WithFilename] to avoid listing file sizes locally
```

Comment:
Consider moving this to the README

Reply:
This is really a design comment, rather than a "TODO" since this code works without needing to be updated in the future. I'd like to leave it in place here (without "TODO") for the reader to understand in context.